subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -820,7 +823,81 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec implements Dire
 
         then:
         executedAndNotSkipped(":customTask")
-        file("output.txt").text == "[100:example][100:example]"
+        file("output.txt").text ==~ /\[100:NestedBean@.*\]\[100:NestedBean@.*\]/
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/23049")
+    @ValidationTestFor(ValidationProblemId.NESTED_TYPE_UNSUITED)
+    def "validation fails for nested #type"() {
+        buildFile << """
+            enum Letter { A, B, C }
+
+            abstract class CustomTask extends DefaultTask {
+                @Nested
+                $type getMy$type() {
+                    return $producer
+                }
+
+                @TaskAction
+                void execute() { }
+            }
+
+            tasks.register("customTask", CustomTask) { }
+        """
+
+        when:
+        expectThatExecutionOptimizationDisabledWarningIsDisplayed(executer,
+            "Type 'CustomTask' property 'my$type' where type of '$className' is unsuited for nested annotation. " +
+                "Reason: Primitive wrapper types and others are unsuited for nested annotation.",
```

Comment:
:x: The reason here should explain the rules we came up with: that nested types must have some annotated properties declared.